### ISSUE TITLE
OC: Bug in handling of 'OEM specified' failsafe for PlatformInfo/Generic/SystemUUID

### ISSUE THREAD

**Issue by Bmju** [acidanthera/bugtracker#1457]
When I boot into my unmodified system and use `dmidecode`, I see the following section:

```
System Information
        Manufacturer: Apple Inc.
        Product Name: MacBookPro10,2
        Version: 1.0
        Serial Number: ************
        UUID: a9bffa99-794f-1e59-9183-**nochange**
        Wake-up Type: Power Switch
        SKU Number: System SKU#
        Family: Mac
```

However, when I boot via OpenCore, with `PlatformInfo/Automatic` = `true` and `PlatformInfo/Generic/SystemUUID` missing (or empty string, I have tried both with the same result) then I get:

```
System Information
        Manufacturer: Apple Inc.
        Product Name: iMac15,1
        Version: 1.0
        Serial Number: ************
        UUID: 99fabfa9-4f79-591e-9183-**nochange**
        Wake-up Type: Power Switch
        SKU Number: System SKU#
        Family: iMac
```

Note the UUID, in which the first three parts are byte-swaps of the OEM original ID; I believe this is a bug and that I should be getting the unmodified original UUID.

**Comment by vit9696** [comment]
Long story short, there is no issue here. Yet the devil is in the details.

Before [SMBIOS 2.6](https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_2.6.0.pdf) there was [no clarification](https://www.dmtf.org/sites/default/files/standards/documents/DSP0134v2.5Final.pdf) how SystemUUID was to be encoded in SMBIOS. And only starting with SMBIOS 2.6 we got this:

> Although RFC4122 recommends network byte order for all fields, the PC industry (including the ACPI, UEFI, and Microsoft specifications) has consistently used little-endian byte encoding for the first three fields: time_low, time_mid, time_hi_and_version. The same encoding, also known as wire format, should also be used for the SMBIOS representation of the UUID.
>
> The UUID `{00112233-4455-6677-8899-AABBCCDDEEFF}` would thus be represented as: `33 22 11 00 55 44 77 66 88 99 AA BB CC DD EE FF`.

As a result the issue is essentially with dmidecode logic. It performs UUID swapping [based on SMBIOS version](https://github.com/acidanthera/dmidecode/blob/1c9b2bb/dmidecode.c#L469-L500). Prior to SMBIOS 2.6.0 it shows as is, and starting with 2.6.0 in little endian. Unlike dmidecode, operating systems, e.g. macOS or Windows, do not seem to care.

It is indeed the case that OpenCore bumps SMBIOS version to 3.x, and if the original version was below 2.6, dmidecode will indeed get confused and start showing different values. For this reason I believe there are three routes:

- Change dmidecode to always do swapping regardless of SMBIOS version.
- Allow setting custom SMBIOS version in OpenCore.
- Do nothing and just ignore the issue.

The last option is personally my preference for the time being. Let me know if you have reasons to choose something else.

**Comment by Bmju** [comment]
@vit9696 

I have checked using `dmidecode` on a clean Ubuntu boot on the same Mac. It reports `UUID: a9bffa99-794f-1e59-****-************`.

Also running my own [nvram var lister .efi](https://github.com/MikeBeaton/macos-boot-helper) during a clean boot, I see:

![Screenshot 2021-01-30 at 11 20 39](https://user-images.githubusercontent.com/25643187/106355175-46882c80-62ee-11eb-9a9d-fca6dac5dcb1.png)

i.e. the Mac itself thinks that its System UUID is `a9bffa99-794f-1e59-****-************`

Thus it seems that the correct System UUID for my Mac is definitely `a9bffa99-794f-1e59-****-************`, so option 1 would be ruled out (it would make `dmidecode` definitely behave incorrectly)

FYI both versions of dmidecode report SMBIOS 2.4 on a non-OC boot.

As reported, when booting through OC set up to reflect the OEM specified value, your `dmidecode` reports `UUID: 99fabfa9-4f79-591e-****-************`. Having Googled around for some time, I still haven't found any other (command line or GUI) way to see what the Mac thinks the value is, from within the booted OS. i.e. to see how the value which the OC SMBIOS 3.2 is providing is being used/understood by anything else which calls it - can you suggest anything? That would seem to be the final additional useful piece of evidence.

However my current thoughts, based on all your additional info, plus the above evidence which I do have, is that perhaps it may still be correct to do the swap which is was proposing in my PR, but only when transferring data from a pre-2.6 SBMIOS into the OC data block. (This would seem to be consistent with how `dmidecode` itself is using that data.)

**Comment by vit9696** [comment]
I am afraid you did not realise what I explained above. There are no options here, there are only suggestions on how to make dmidecode happy if that is really desired. Because the **only** thing that swaps your UUID is dmidecode, and it does that when giving you human-readable byte sequence representation. It does that based on the SMBIOS version, which for OpenCore sets to 3.3 and without OpenCore for you is 2.4.

macOS never swaps, and it always reads your SMBIOS UUID according to GUID RFC, not according to SMBIOS 2.6+ spec suggestion. For all operating systems it is the binary sequence that matters, they do not "think" in the UUID format. For them it is just 16 bytes in a certain order. Doing any kind of changes to these bytes will break your system, e.g. Windows activation if the amount of parameter changes exceeds the tolerance level.

Thus the important thing here is to give the same binary value of UUID regardless of what dmidecode (incorrectly) thinks. You can do a binary SMBIOS dump via I/O Registry Explorer and find the relevant bytes via any hex editor. You will see that these are both same sequences with and without OpenCore.

**Comment by vit9696** [comment]
I updated dmidecode to print in all three formats, you should now see the following unambiguous values:

```
plumeria:dmidecode user$ ./dmidecode | grep UUID
	UUID: {49 98 06 fe ...} (RAW)
	UUID: fe069849-... (SMBIOS 2.6+, preferred)
	UUID: 499806fe-... (RFC 4122, Apple)
```

The `preferred` part signalises the SMBIOS version. You should always rely on comparing `RAW` line **only** when determining whether the UUID is correctly set.

**Comment by Bmju** [comment]
TL/DR: I agree. It all works. Many thanks for the changes.

----

I verified, using the Windows command `wmic csproduct get UUID` on a bootcamp boot on the same Mac, both in and out of OC, that everything is indeed working correctly and as you say. That Windows command also shows the correct value which I need to set for the explicit UUID in OC (it is `99fabfa9...`, the opposite of what I thought).

It was very confusing because the previous `dmidecode` (when run in a non-OC boot) showed me a value which I assumed was right, but which was in fact wrong (not what Windows shows, not what I should set to preserve the value). Then when I set that wrong value, booted and (again) used the previous `dmidecode` to check things it (correctly, this time!) showed me that that incorrect value was set ... thereby leading me to assume that it was correct! 🤯

I honestly doubt I would have been the only person confused by this! The updated tooling (and the explanation above) are a huge help. Thanks!

**Comment by Bmju** [comment]
@vit9696 Apologies, I know this is now getting boring. I am not trying to track down bugs (or non-bug problems) for the sake of it.

The current `macserial` tool behaves in the same way as the previous `dmidecode` tool (which is also behaving like the Ubuntu `dmidecode` and the cavaliercoder `dmidecode` for Mac).

Namely, I run any of them on my Mac not booted via OC (under Ubuntu or macOS as appropriate). They report `System ID: A9BFFA99-794F-1E59-9183-3EE79D67BB4F`. I put this value into my `Generic` OC config and re-boot into OC. I run the respective tool again. It reports `System ID: A9BFFA99-794F-1E59-9183-3EE79D67BB4F`. _It looks like everything has worked!_ But it hasn't. As your new `dmidecode` shows, the byte order has actually changed.

(It's almost like the Un*x-world understanding of the UUID is that it is expected to be stored in a different byte-order in different SBMIOS versions, and the Windows-world understanding is that the UUID is always stored in a fixed order regardless of SMBIOS version.)

**Comment by vit9696** [comment]
Oh dear… Well, let me explain another ugly story.

First of all, System UUID is actually present in several places in macOS. First place is SMBIOS and second place is DataHub. You can observe the `SMBIOS` value via `dmidecode`, and DataHub value is accessible from IORegistryExplorer `IODeviceTree:/efi/platform - system-id`.

I am not sure why, but for DataHub you must perform byte order swap. This is at least what we observed in the Macs we had access to. I.e. in SMBIOS you have 11 22 33 44, and in DataHub you have 44 33 22 11 (all RAW values).

macserial prints from DataHub without byte swapping 1st2nd3rd4th-5th6th… Hope it helps.

**Comment by Bmju** [comment]
Yes, it helps.

Not sure how much I'd been putting in edits, but to briefly repeat with a new bit that I was in the middle of adding as an edit (didn't realise you'd be reading at this time of night, or is it morning now?!):

It's almost like the Un*x-world understanding of the UUID is that it is expected to be stored in a different byte-order in different SBMIOS (>= 2.6) versions, while the Windows-world understanding is that the UUID is always stored in a fixed order regardless of SMBIOS version.

And if that is right, then it's not really possible to cross from below to above or equal to 2.6 SMBIOS version and have both Windows and Mac think that the System UUID has not changed - but since we care more about Windows (activation) we go by byte order.

I don't know if you agree with that? It doesn't really matter, I agree nothing (else) needs changing at this point!

**Comment by Bmju** [comment]
TL/DR: I'm very, very sorry, but I am going to suggest again that something 'needs' to be changed; or, rather, still could helpfully, usefully, be changed.

IF I'm right that to represent the same system uuid on macOS, on pre- vs post- 2.6 SMBIOS, requires a different byte order (thereby necessarily and unfortunately making the very same set of bytes look like a different system id to Windows), then:

1. When using the OEM specified system uuid from a pre-2.6 original SBMBIOS structure, OC still should swap the bytes where I suggested _to be correct for macOS_ (and this will be wrong for Windows; whereas not swapping them in this case is right for Windows - which matters, for Windows authorisation - but wrong for macOS)
1. It is worth very briefly explaining a relevant detail of my use case: I normally (except for these tests) boot via OC only for macOS (not for Windows or Ubuntu), and do so only occasionally to get Big Sur updates (I use the Mac boot picker for everything else the rest of the time); therefore I would actually (although macOS doesn't seem to care) prefer the above swap for my case, based on my current understanding of what is happening (I agree that I need, and lack, incontroversial evidence that macOS a) cares about this at all, and b) really is reading the raw SMBIOS data the way I'm suggesting; I don't have any ... except for the consistent output of all the different dmi tools, including your own earlier versions, on Windows and on macOS)
1. Finally, if I am right, then it still would actually be very useful to have an optional config flag (default `false`) in the `<Generic>` section to request OC to swap these bytes, something like `<SwapSystemUUID>`; this would be sensible to use on machines which a) natively have a pre-2.6 SMBIOS, and b) on which you are not booting Windows via OC (or else, have not previously booted and authorised Windows not via OC even if you are planning to do so after installing OC, because in that case you don't care what value Windows _previously would have read, but never actually did so_).

The only other alternatives (to support this tiny corner case in a completely gold plated and over-complicated manner) I would agree involve making OC support reporting more than one SMBIOS version (I realise there are very likely implications beyond just 'reporting a different number', i.e that the behaviour probably has to change to match?). I am suggesting that this is too much to consider doing for this tiny reason, hence the above suggestion instead. But for completeness, this could include:
  1. Allow user to manually specify the reported SMBIOS version number (as you already suggested as one option)
  2. Make OC always report the original OEM specified number (or, at least, allow to do so on a flag)
  3. Make OC always use (say) SMBIOS version 2.5 for all original versions < 2.6, and 3.x otherwise

**Comment by vit9696** [comment]
To be honest, I am starting to miss the point.

- macOS and Windows read UUIDs consistently regardless of SMBIOS versions.
- As long as UUID is not specified in OC, it is left unchanged.
- Old dmidecode was confused, but w/e really, we pushed a fix for it.

**Comment by vit9696** [comment]
I think I get what your problem is. OpenCore reads all the UUIDs specified in the configuration it Little Endian (swapped) order as this is the order generally used in UEFI. I.e. 11223344-5566... in OC config.plist becomes {44 33 22 11 66 55…} in SMBIOS and NVRAM system-id. DataHub System UUID is an exception, and it is written in Big Endian (RFC 4122) — {11 22 33 44 55 66…}.

This leaves us in a situation where OpenCore does not allow for the same byte values to be written in all the three places. But instead the first two will always be different from the third. This is not how Apple hardware works, where all the UUIDs apparently are in RFC 4122.

I wonder whether the swapping is missing for SMBIOS (and nvram) due to that messed up swapping dmidecode developers introduced following SMBIOS 2.6 clarification. Unsure whether we need to resolve this now and how, but I do agree that this is a problem at least.

**Comment by MikeBeaton** [comment]
I'm actually arguing that Apple's software would send a different authorisation hash over the wire to its servers (e.g. for iMessage, FaceTime) for the same bytes in the SMBIOS block depending on whether the reported SMBIOS version is below 2.6 or not. I felt this must be true because *all* the tools (I've now found iMessageDebug2, too) are consistently reading with UUID like this. But basically I'm wrong about that, if that's not what Apple's software does!

I'm trying to do some more research on this. If I find anything new, I'll try to summarise it, briefly!

**Comment by vit9696** [comment]
Both macserial, iMessageDebug2, and others read UUID from DataHub. It may or may not be different from SMBIOS. Your problem is when you set this UUID, it is encoded differently in SMBIOS.

**Comment by vit9696** [comment]
Please test the version in https://github.com/acidanthera/OpenCorePkg/pull/187. I believe it fully explains and resolves your problem.

Filling the log with zeroes is intentional. All the unused bytes in the log are zeroes to maintain constant file size avoiding file system corruption due to driver flaws. This is what is implied by "Log writing is attempted in the safest manner and thus" in Configuration.pdf actually.

**Comment by Bmju** [comment]
Will test, thank you. "Eres una maquina", as they say in Spain - you produce this stuff like a machine - in the best sense!
